File: python-student/Student.py
import grpc
import uuid
import logging
import Classroom_pb2_grpc
from typing import List

from Classroom_pb2 import Question

logger = logging.getLogger('Student')
logger.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
ch.setFormatter(logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s'))
logger.addHandler(ch)


class Student:
    def __init__(self) -> None:
        super().__init__()
        channel = grpc.insecure_channel('localhost:50051')  # no ssl
        self.asyncProfessor = Classroom_pb2_grpc.ProfessorServiceStub(channel)

    def ask_simple_question(self, question: str):
        logger.info("[Python Student]: {}".format(question))

        question_proto = Question(id=str(uuid.uuid1()), text=question)

        answer = self.asyncProfessor.askSimpleQuestion(question_proto)

        logger.info("[Professor]: {}".format(answer.text))

        logger.info("[Python Student]: Thank you professor.")

    def ask_multiple_simple_questions(self, questions: List[str]):
        logger.info("[Python Student]: {}".format(' | '.join(questions)))

        question_protos = (Question(id=str(uuid.uuid1()), text=question) for question in questions)

        answer = self.asyncProfessor.askMultipleSimpleQuestions(question_protos)

        logger.info("[Professor]: {}".format(answer.text))

        logger.info("[Python Student]: Thank you professor.")

    def ask_complex_question(self, question: str):
        logger.info("[Python Student]: {}".format(question))

        question_proto = Question(id=str(uuid.uuid1()), text=question)

        answers = self.asyncProfessor.askComplexQuestion(question_proto)

        for answer in answers:
            logger.info("[Professor]: {}".format(answer.text))

        logger.info("[Python Student]: Thank you professor.")

    def ask_multiple_complex_questions(self, questions: List[str]):
        logger.info("[Python Student]: {}".format(' | '.join(questions)))

        question_protos = (Question(id=str(uuid.uuid1()), text=question) for question in questions)

        answers = self.asyncProfessor.askMultipleComplexQuestions(question_protos)

        for answer in answers:
            logger.info("[Professor]: {}".format(answer.text))

        logger.info("[Python Student]: Thank you professor.")

File: java-student/src/main/java/at/jku/isse/psma/tutorial/grpc/Student.java
package at.jku.isse.psma.tutorial.grpc;

import io.grpc.Channel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.UUID;

public class Student {

    private static final Logger logger = LoggerFactory.getLogger(Student.class);

    private final ProfessorServiceGrpc.ProfessorServiceStub asyncProfessor;

    public Student() {
        Channel channel = ManagedChannelBuilder.forAddress("localhost", 50051).build();
        asyncProfessor = ProfessorServiceGrpc.newStub(channel);
    }

    private class LoggerStreamObserver implements StreamObserver<Answer> {
        @Override
        public void onNext(final Answer value) {
            logger.info("Professor: {}", value.getText());
        }

        @Override
        public void onError(final Throwable t) {
            logger.error("Something disturbed the classroom.", t);
        }

        @Override
        public void onCompleted() {
            logger.info("Java Student: Thank you professor.");
        }
    }

    private Question newQuestion(String question) {
        return Question.newBuilder()
                .setId(UUID.randomUUID().toString())
                .setText(question)
                .build();
    }

    public void askSimpleQuestion(String question) {
        logger.info("Asking the professor a simple question.");

        Question questionProto = newQuestion(question);

        asyncProfessor.askQuestionWithAnswer(questionProto, new LoggerStreamObserver());

        logger.info("Finished asking.");
    }

    public void askMultipleSimpleQuestions(List<String> questions) {
        logger.info("Asking the professor multiple simple questions.");

        StreamObserver<Question> requestStreamObserver = asyncProfessor.askQuestionsWithAnswer(new LoggerStreamObserver());

        questions.stream()
                .map(this::newQuestion)
                .forEach(requestStreamObserver::onNext);

        requestStreamObserver.onCompleted();

        logger.info("Finished asking.");
    }

    public void askComplexQuestion(String question) {
        logger.info("Asking the professor a complex question.");

        Question questionProto = newQuestion(question);

        asyncProfessor.askQuestionWithAnswer(questionProto, new LoggerStreamObserver());

        logger.info("Finished asking.");
    }

    public void askMultipleComplexQuestions(List<String> questions) {
        logger.info("Asking the professor multiple complex questions.");

        StreamObserver<Question> requestStreamObserver = asyncProfessor.askQuestionsWithAnswer(new LoggerStreamObserver());

        questions.stream()
                .map(this::newQuestion)
                .forEach(requestStreamObserver::onNext);

        requestStreamObserver.onCompleted();

        logger.info("Finished asking.");
    }
}
